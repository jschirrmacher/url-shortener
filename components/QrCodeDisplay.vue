<script setup lang="ts">
interface Props {
  shortCode: string
  size?: number
  showDownload?: boolean
}

const props = withDefaults(defineProps<Props>(), {
  size: 200,
  showDownload: true,
})

const qrCodeUrl = computed(() => `/api/qr/${props.shortCode}?size=${props.size}`)
const downloadUrl = computed(() => `/api/qr/${props.shortCode}?size=${props.size}&download=true`)
const svgDownloadUrl = computed(() => `/api/qr/${props.shortCode}?format=svg&size=${props.size}&download=true`)

const isLoading = ref(true)
const hasError = ref(false)
const errorMessage = ref("")
const isClient = ref(false)

// Check if we're on client side
onMounted(() => {
  isClient.value = true
})

function handleImageLoad() {
  isLoading.value = false
  hasError.value = false
  errorMessage.value = ""
}

function handleImageError() {
  isLoading.value = false
  hasError.value = true
  errorMessage.value = "QR-Code konnte nicht geladen werden"
}

async function downloadQrCode(format: "png" | "svg") {
  try {
    const url = format === "svg" ? svgDownloadUrl.value : downloadUrl.value
    const filename = `qr-${props.shortCode}.${format}`
    const link = document.createElement("a")
    link.href = url
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Download failed:", error)
  }
}

// Reset loading state when shortCode changes
watch(
  () => props.shortCode,
  () => {
    isLoading.value = true
    hasError.value = false
    errorMessage.value = ""
  },
)
</script>

<template>
  <div class="qr-code-container">
    <div class="qr-code-wrapper" :style="{ width: `${size}px`, height: `${size}px` }">
      <!-- Show loading initially -->
      <div v-if="!isClient || isLoading" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
        <div class="text-center">
          <div class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto" />
          <p class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
        </div>
      </div>

      <!-- Show error state -->
      <div
        v-else-if="hasError"
        class="flex items-center justify-center w-full h-full bg-red-50 border-2 border-red-200 rounded-lg"
      >
        <div class="text-center">
          <svg class="mx-auto h-8 w-8 text-red-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.732-.833-2.5 0L4.268 19.5c-.77.833.192 2.5 1.732 2.5z"
            />
          </svg>
          <p class="text-sm text-red-600 mt-1">{{ errorMessage }}</p>
          <p class="text-xs text-gray-500 mt-1">URL: {{ qrCodeUrl }}</p>
        </div>
      </div>

      <ClientOnly>
        <img
          :src="qrCodeUrl"
          :alt="`QR-Code fÃ¼r ${shortCode}`"
          class="w-full h-full object-contain rounded-lg border border-gray-200"
          @load="handleImageLoad"
          @error="handleImageError"
        >
      </ClientOnly>
    </div>

    <!-- Download Buttons - only show when QR code is loaded -->
    <div v-if="isClient && showDownload && !isLoading && !hasError" class="mt-4 flex gap-2 justify-center">
      <button
        class="inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
        @click="downloadQrCode('png')"
      >
        <svg class="h-4 w-4 mr-2" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M12 10v6m0 0l-3-3m3 3l3-3m2 8H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
          />
        </svg>
        PNG herunterladen
      </button>

      <button
        class="inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
        @click="downloadQrCode('svg')"
      >
        <svg class="h-4 w-4 mr-2" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M12 10v6m0 0l-3-3m3 3l3-3m2 8H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
          />
        </svg>
        SVG herunterladen
      </button>
    </div>
  </div>
</template>

<style scoped>
.qr-code-container {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.qr-code-wrapper {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
}
</style>
