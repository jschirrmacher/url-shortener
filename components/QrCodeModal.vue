<script setup lang="ts">
interface Props {
  shortCode: string
  shortUrl: string
  originalUrl: string
  isOpen: boolean
}

interface Emits {
  (e: "close"): void
}

const props = defineProps<Props>()
const emit = defineEmits<Emits>()

const selectedSize = ref(200)
const availableSizes = [
  { value: 150, label: "Klein (150px)" },
  { value: 200, label: "Mittel (200px)" },
  { value: 300, label: "Groß (300px)" },
  { value: 400, label: "Extra Groß (400px)" }
]

function closeModal() {
  emit("close")
}

async function copyShortUrl() {
  try {
    await navigator.clipboard.writeText(props.shortUrl)
    // Could add a toast notification here
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error('Copy failed:', error)
  }
}
</script>

<template>
  <div
    v-if="isOpen"
    class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4"
    @click.self="closeModal"
  >
    <div class="bg-white rounded-lg shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto">
      <div class="px-6 py-4 border-b border-gray-200">
        <div class="flex items-center justify-between">
          <h3 class="text-lg font-semibold text-gray-900">
            QR-Code für Kurzlink
          </h3>
          <button
            class="text-gray-400 hover:text-gray-600 transition-colors"
            @click="closeModal"
          >
            <svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12" />
            </svg>
          </button>
        </div>
      </div>

      <div class="px-6 py-4">
        <!-- URL Info -->
        <div class="mb-6">
          <div class="mb-2">
            <label class="block text-sm font-medium text-gray-700 mb-1">
              Kurzlink:
            </label>
            <div class="flex items-center gap-2">
              <code class="flex-1 px-3 py-2 bg-gray-50 border border-gray-300 rounded-md text-sm font-mono">
                {{ shortUrl }}
              </code>
              <button
                class="px-3 py-2 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
                title="Link kopieren"
                @click="copyShortUrl"
              >
                <svg class="h-4 w-4" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                  <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M8 16H6a2 2 0 01-2-2V6a2 2 0 012-2h8a2 2 0 012 2v2m-6 12h8a2 2 0 002-2v-8a2 2 0 00-2-2h-8a2 2 0 00-2 2v8a2 2 0 002 2z" />
                </svg>
              </button>
            </div>
          </div>
          
          <div>
            <label class="block text-sm font-medium text-gray-700 mb-1">
              Ziel-URL:
            </label>
            <p class="text-sm text-gray-600 break-all">{{ originalUrl }}</p>
          </div>
        </div>

        <!-- Size Selection -->
        <div class="mb-6">
          <label class="block text-sm font-medium text-gray-700 mb-2">
            QR-Code Größe:
          </label>
          <select
            v-model="selectedSize"
            class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent"
          >
            <option
              v-for="size in availableSizes"
              :key="size.value"
              :value="size.value"
            >
              {{ size.label }}
            </option>
          </select>
        </div>

        <!-- QR Code Display -->
        <div class="flex justify-center">
          <QrCodeDisplay
            :short-code="shortCode"
            :size="selectedSize"
            :show-download="true"
          />
        </div>
      </div>

      <div class="px-6 py-4 border-t border-gray-200 bg-gray-50">
        <div class="flex justify-end">
          <button
            class="px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
            @click="closeModal"
          >
            Schließen
          </button>
        </div>
      </div>
    </div>
  </div>
</template>
