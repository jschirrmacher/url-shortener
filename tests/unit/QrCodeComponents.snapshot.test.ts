/**
 * @vitest-environment happy-dom
 */
import { describe, it, expect } from "vitest"
import { mount } from "@vue/test-utils"
import QrCodeDisplay from "~/components/QrCodeDisplay.vue"
import QrCodeModal from "~/components/QrCodeModal.vue"

describe("QR Code Components Snapshots", () => {
  describe("QrCodeDisplay", () => {
    it("should render with default props", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "abc123",
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with custom size", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "abc123",
          size: 300,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render without download buttons", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "abc123",
          showDownload: false,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with all props", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "test123",
          size: 250,
          showDownload: true,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe("QrCodeModal", () => {
    const defaultProps = {
      shortCode: "abc123",
      shortUrl: "https://example.com/abc123",
      originalUrl: "https://target.com",
      isOpen: true,
    }

    it("should render when open", () => {
      const wrapper = mount(QrCodeModal, {
        props: defaultProps,
        global: {
          stubs: {
            QrCodeDisplay: {
              template: '<div class="qr-display-stub">QR Display</div>',
            },
          },
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should not render when closed", () => {
      const wrapper = mount(QrCodeModal, {
        props: {
          ...defaultProps,
          isOpen: false,
        },
        global: {
          stubs: {
            QrCodeDisplay: {
              template: '<div class="qr-display-stub">QR Display</div>',
            },
          },
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with long URLs", () => {
      const wrapper = mount(QrCodeModal, {
        props: {
          ...defaultProps,
          shortUrl: "https://very-long-domain-name.example.com/very/long/path/abc123",
          originalUrl:
            "https://another-very-long-domain-name.example.com/with/many/path/segments/and/parameters?param1=value1&param2=value2",
        },
        global: {
          stubs: {
            QrCodeDisplay: {
              template: '<div class="qr-display-stub">QR Display</div>',
            },
          },
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should render with special characters in URLs", () => {
      const wrapper = mount(QrCodeModal, {
        props: {
          ...defaultProps,
          shortUrl: "https://example.com/special123",
          originalUrl: "https://target.com/path?param=value&other=test#section",
        },
        global: {
          stubs: {
            QrCodeDisplay: {
              template: '<div class="qr-display-stub">QR Display</div>',
            },
          },
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe("Component Structure", () => {
    it("should maintain consistent QrCodeDisplay structure", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "structure",
        },
      })

      // Test key structural elements
      expect(wrapper.find(".qr-code-container").exists()).toBe(true)
      expect(wrapper.find(".qr-code-wrapper").exists()).toBe(true)
      expect(wrapper.text()).toContain("QR-Code für: structure")
    })

    it("should maintain consistent QrCodeModal structure", () => {
      const wrapper = mount(QrCodeModal, {
        props: {
          shortCode: "modal123",
          shortUrl: "https://example.com/modal123",
          originalUrl: "https://target.com",
          isOpen: true,
        },
        global: {
          stubs: {
            QrCodeDisplay: {
              template: '<div class="qr-display-stub">QR Display</div>',
            },
          },
        },
      })

      // Test key structural elements
      expect(wrapper.find(".fixed.inset-0").exists()).toBe(true)
      expect(wrapper.find(".bg-white.rounded-lg").exists()).toBe(true)
      expect(wrapper.find("select").exists()).toBe(true)
      expect(wrapper.text()).toContain("QR-Code für Kurzlink")
      expect(wrapper.text()).toContain("modal123")
    })
  })

  describe("Props Validation", () => {
    it("should handle empty shortCode gracefully", () => {
      const wrapper = mount(QrCodeDisplay, {
        props: {
          shortCode: "",
        },
      })

      expect(wrapper.exists()).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should handle minimal modal props", () => {
      const wrapper = mount(QrCodeModal, {
        props: {
          shortCode: "",
          shortUrl: "",
          originalUrl: "",
          isOpen: true,
        },
        global: {
          stubs: {
            QrCodeDisplay: {
              template: '<div class="qr-display-stub">QR Display</div>',
            },
          },
        },
      })

      expect(wrapper.exists()).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
