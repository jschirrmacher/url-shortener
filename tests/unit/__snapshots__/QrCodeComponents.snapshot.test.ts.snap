// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QR Code Components Snapshots > Props Validation > should handle empty shortCode gracefully 1`] = `
"<div data-v-9d2b417e="" class="qr-code-container">
  <div data-v-9d2b417e="" class="qr-code-wrapper" style="width: 200px; height: 200px;">
    <!-- Show loading initially -->
    <div data-v-9d2b417e="" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
      <div data-v-9d2b417e="" class="text-center">
        <div data-v-9d2b417e="" class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto"></div>
        <p data-v-9d2b417e="" class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
      </div>
    </div>
  </div><!-- Download Buttons - only show when QR code is loaded -->
  <!--v-if-->
  <!-- Info Text -->
  <p data-v-9d2b417e="" class="text-xs text-gray-500 text-center mt-2"> QR-Code für: <span data-v-9d2b417e="" class="font-mono"></span></p>
</div>"
`;

exports[`QR Code Components Snapshots > Props Validation > should handle minimal modal props 1`] = `
"<div class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4">
  <div class="bg-white rounded-lg shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto">
    <div class="px-6 py-4 border-b border-gray-200">
      <div class="flex items-center justify-between">
        <h3 class="text-lg font-semibold text-gray-900"> QR-Code für Kurzlink </h3><button class="text-gray-400 hover:text-gray-600 transition-colors"><svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"></path>
          </svg></button>
      </div>
    </div>
    <div class="px-6 py-4">
      <!-- URL Info -->
      <div class="mb-6">
        <div class="mb-2"><label class="block text-sm font-medium text-gray-700 mb-1"> Kurzlink: </label>
          <div class="flex items-center gap-2"><code class="flex-1 px-3 py-2 bg-gray-50 border border-gray-300 rounded-md text-sm font-mono"></code><button class="px-3 py-2 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2" title="Link kopieren"><svg class="h-4 w-4" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M8 16H6a2 2 0 01-2-2V6a2 2 0 012-2h8a2 2 0 012 2v2m-6 12h8a2 2 0 002-2v-8a2 2 0 00-2-2h-8a2 2 0 00-2 2v8a2 2 0 002 2z"></path>
              </svg></button></div>
        </div>
        <div><label class="block text-sm font-medium text-gray-700 mb-1"> Ziel-URL: </label>
          <p class="text-sm text-gray-600 break-all"></p>
        </div>
      </div><!-- Size Selection -->
      <div class="mb-6"><label class="block text-sm font-medium text-gray-700 mb-2"> QR-Code Größe: </label><select class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent">
          <option value="150">Klein (150px)</option>
          <option value="200">Mittel (200px)</option>
          <option value="300">Groß (300px)</option>
          <option value="400">Extra Groß (400px)</option>
        </select></div><!-- QR Code Display -->
      <div class="flex justify-center">
        <div class="qr-display-stub" short-code="" size="[object Object]" show-download="true">QR Display</div>
      </div>
    </div>
    <div class="px-6 py-4 border-t border-gray-200 bg-gray-50">
      <div class="flex justify-end"><button class="px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"> Schließen </button></div>
    </div>
  </div>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeDisplay > should render with all props 1`] = `
"<div data-v-9d2b417e="" class="qr-code-container">
  <div data-v-9d2b417e="" class="qr-code-wrapper" style="width: 250px; height: 250px;">
    <!-- Show loading initially -->
    <div data-v-9d2b417e="" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
      <div data-v-9d2b417e="" class="text-center">
        <div data-v-9d2b417e="" class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto"></div>
        <p data-v-9d2b417e="" class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
      </div>
    </div>
  </div><!-- Download Buttons - only show when QR code is loaded -->
  <!--v-if-->
  <!-- Info Text -->
  <p data-v-9d2b417e="" class="text-xs text-gray-500 text-center mt-2"> QR-Code für: <span data-v-9d2b417e="" class="font-mono">test123</span></p>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeDisplay > should render with custom size 1`] = `
"<div data-v-9d2b417e="" class="qr-code-container">
  <div data-v-9d2b417e="" class="qr-code-wrapper" style="width: 300px; height: 300px;">
    <!-- Show loading initially -->
    <div data-v-9d2b417e="" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
      <div data-v-9d2b417e="" class="text-center">
        <div data-v-9d2b417e="" class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto"></div>
        <p data-v-9d2b417e="" class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
      </div>
    </div>
  </div><!-- Download Buttons - only show when QR code is loaded -->
  <!--v-if-->
  <!-- Info Text -->
  <p data-v-9d2b417e="" class="text-xs text-gray-500 text-center mt-2"> QR-Code für: <span data-v-9d2b417e="" class="font-mono">abc123</span></p>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeDisplay > should render with default props 1`] = `
"<div data-v-9d2b417e="" class="qr-code-container">
  <div data-v-9d2b417e="" class="qr-code-wrapper" style="width: 200px; height: 200px;">
    <!-- Show loading initially -->
    <div data-v-9d2b417e="" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
      <div data-v-9d2b417e="" class="text-center">
        <div data-v-9d2b417e="" class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto"></div>
        <p data-v-9d2b417e="" class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
      </div>
    </div>
  </div><!-- Download Buttons - only show when QR code is loaded -->
  <!--v-if-->
  <!-- Info Text -->
  <p data-v-9d2b417e="" class="text-xs text-gray-500 text-center mt-2"> QR-Code für: <span data-v-9d2b417e="" class="font-mono">abc123</span></p>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeDisplay > should render without download buttons 1`] = `
"<div data-v-9d2b417e="" class="qr-code-container">
  <div data-v-9d2b417e="" class="qr-code-wrapper" style="width: 200px; height: 200px;">
    <!-- Show loading initially -->
    <div data-v-9d2b417e="" class="flex items-center justify-center w-full h-full bg-gray-100 rounded-lg">
      <div data-v-9d2b417e="" class="text-center">
        <div data-v-9d2b417e="" class="animate-spin rounded-full h-8 w-8 border-b-2 border-blue-600 mx-auto"></div>
        <p data-v-9d2b417e="" class="text-xs text-gray-500 mt-2">QR-Code wird geladen...</p>
      </div>
    </div>
  </div><!-- Download Buttons - only show when QR code is loaded -->
  <!--v-if-->
  <!-- Info Text -->
  <p data-v-9d2b417e="" class="text-xs text-gray-500 text-center mt-2"> QR-Code für: <span data-v-9d2b417e="" class="font-mono">abc123</span></p>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeModal > should not render when closed 1`] = `"<!--v-if-->"`;

exports[`QR Code Components Snapshots > QrCodeModal > should render when open 1`] = `
"<div class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4">
  <div class="bg-white rounded-lg shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto">
    <div class="px-6 py-4 border-b border-gray-200">
      <div class="flex items-center justify-between">
        <h3 class="text-lg font-semibold text-gray-900"> QR-Code für Kurzlink </h3><button class="text-gray-400 hover:text-gray-600 transition-colors"><svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"></path>
          </svg></button>
      </div>
    </div>
    <div class="px-6 py-4">
      <!-- URL Info -->
      <div class="mb-6">
        <div class="mb-2"><label class="block text-sm font-medium text-gray-700 mb-1"> Kurzlink: </label>
          <div class="flex items-center gap-2"><code class="flex-1 px-3 py-2 bg-gray-50 border border-gray-300 rounded-md text-sm font-mono">https://example.com/abc123</code><button class="px-3 py-2 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2" title="Link kopieren"><svg class="h-4 w-4" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M8 16H6a2 2 0 01-2-2V6a2 2 0 012-2h8a2 2 0 012 2v2m-6 12h8a2 2 0 002-2v-8a2 2 0 00-2-2h-8a2 2 0 00-2 2v8a2 2 0 002 2z"></path>
              </svg></button></div>
        </div>
        <div><label class="block text-sm font-medium text-gray-700 mb-1"> Ziel-URL: </label>
          <p class="text-sm text-gray-600 break-all">https://target.com</p>
        </div>
      </div><!-- Size Selection -->
      <div class="mb-6"><label class="block text-sm font-medium text-gray-700 mb-2"> QR-Code Größe: </label><select class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent">
          <option value="150">Klein (150px)</option>
          <option value="200">Mittel (200px)</option>
          <option value="300">Groß (300px)</option>
          <option value="400">Extra Groß (400px)</option>
        </select></div><!-- QR Code Display -->
      <div class="flex justify-center">
        <div class="qr-display-stub" short-code="abc123" size="[object Object]" show-download="true">QR Display</div>
      </div>
    </div>
    <div class="px-6 py-4 border-t border-gray-200 bg-gray-50">
      <div class="flex justify-end"><button class="px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"> Schließen </button></div>
    </div>
  </div>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeModal > should render with long URLs 1`] = `
"<div class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4">
  <div class="bg-white rounded-lg shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto">
    <div class="px-6 py-4 border-b border-gray-200">
      <div class="flex items-center justify-between">
        <h3 class="text-lg font-semibold text-gray-900"> QR-Code für Kurzlink </h3><button class="text-gray-400 hover:text-gray-600 transition-colors"><svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"></path>
          </svg></button>
      </div>
    </div>
    <div class="px-6 py-4">
      <!-- URL Info -->
      <div class="mb-6">
        <div class="mb-2"><label class="block text-sm font-medium text-gray-700 mb-1"> Kurzlink: </label>
          <div class="flex items-center gap-2"><code class="flex-1 px-3 py-2 bg-gray-50 border border-gray-300 rounded-md text-sm font-mono">https://very-long-domain-name.example.com/very/long/path/abc123</code><button class="px-3 py-2 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2" title="Link kopieren"><svg class="h-4 w-4" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M8 16H6a2 2 0 01-2-2V6a2 2 0 012-2h8a2 2 0 012 2v2m-6 12h8a2 2 0 002-2v-8a2 2 0 00-2-2h-8a2 2 0 00-2 2v8a2 2 0 002 2z"></path>
              </svg></button></div>
        </div>
        <div><label class="block text-sm font-medium text-gray-700 mb-1"> Ziel-URL: </label>
          <p class="text-sm text-gray-600 break-all">https://another-very-long-domain-name.example.com/with/many/path/segments/and/parameters?param1=value1&amp;param2=value2</p>
        </div>
      </div><!-- Size Selection -->
      <div class="mb-6"><label class="block text-sm font-medium text-gray-700 mb-2"> QR-Code Größe: </label><select class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent">
          <option value="150">Klein (150px)</option>
          <option value="200">Mittel (200px)</option>
          <option value="300">Groß (300px)</option>
          <option value="400">Extra Groß (400px)</option>
        </select></div><!-- QR Code Display -->
      <div class="flex justify-center">
        <div class="qr-display-stub" short-code="abc123" size="[object Object]" show-download="true">QR Display</div>
      </div>
    </div>
    <div class="px-6 py-4 border-t border-gray-200 bg-gray-50">
      <div class="flex justify-end"><button class="px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"> Schließen </button></div>
    </div>
  </div>
</div>"
`;

exports[`QR Code Components Snapshots > QrCodeModal > should render with special characters in URLs 1`] = `
"<div class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50 p-4">
  <div class="bg-white rounded-lg shadow-xl max-w-md w-full max-h-[90vh] overflow-y-auto">
    <div class="px-6 py-4 border-b border-gray-200">
      <div class="flex items-center justify-between">
        <h3 class="text-lg font-semibold text-gray-900"> QR-Code für Kurzlink </h3><button class="text-gray-400 hover:text-gray-600 transition-colors"><svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"></path>
          </svg></button>
      </div>
    </div>
    <div class="px-6 py-4">
      <!-- URL Info -->
      <div class="mb-6">
        <div class="mb-2"><label class="block text-sm font-medium text-gray-700 mb-1"> Kurzlink: </label>
          <div class="flex items-center gap-2"><code class="flex-1 px-3 py-2 bg-gray-50 border border-gray-300 rounded-md text-sm font-mono">https://example.com/special123</code><button class="px-3 py-2 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2" title="Link kopieren"><svg class="h-4 w-4" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M8 16H6a2 2 0 01-2-2V6a2 2 0 012-2h8a2 2 0 012 2v2m-6 12h8a2 2 0 002-2v-8a2 2 0 00-2-2h-8a2 2 0 00-2 2v8a2 2 0 002 2z"></path>
              </svg></button></div>
        </div>
        <div><label class="block text-sm font-medium text-gray-700 mb-1"> Ziel-URL: </label>
          <p class="text-sm text-gray-600 break-all">https://target.com/path?param=value&amp;other=test#section</p>
        </div>
      </div><!-- Size Selection -->
      <div class="mb-6"><label class="block text-sm font-medium text-gray-700 mb-2"> QR-Code Größe: </label><select class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent">
          <option value="150">Klein (150px)</option>
          <option value="200">Mittel (200px)</option>
          <option value="300">Groß (300px)</option>
          <option value="400">Extra Groß (400px)</option>
        </select></div><!-- QR Code Display -->
      <div class="flex justify-center">
        <div class="qr-display-stub" short-code="abc123" size="[object Object]" show-download="true">QR Display</div>
      </div>
    </div>
    <div class="px-6 py-4 border-t border-gray-200 bg-gray-50">
      <div class="flex justify-end"><button class="px-4 py-2 text-sm font-medium text-gray-700 bg-white border border-gray-300 rounded-md hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"> Schließen </button></div>
    </div>
  </div>
</div>"
`;
